### PR TITLE
Ensure leader services are closed before starting new ones

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
Current leader transitions in tpu don't ensure leader services are actually shut down before starting up new ones, could potentially cause data races in the future.

#### Summary of Changes
Make sure leader services are actually shut down before starting new ones

Fixes #

### CODE REVIEW

**Comment by garious** [comment]
That doesn't seem very elegant.  You should be able to send an exit signal to *only* fetch_stage and cluster_info_vote_listener. Once its channels are closed, the others should finish up and exit themselves.

**Comment by carllin** [comment]
@garious, I dont think this PR touches the mechanism by which those stages close. This is a change to force leader transitions in the TPU to confirm that those stages are actually shut down before starting up new versions of those stages.

**Comment by garious** [comment]
@carllin, if those stages aren't shutting down, we need to fix the mechanism by which they close, no? This change hits them with a sledgehammer. Have you looked into what this approach would do to features like the one mvines recently added where remaining transactions are forwarded to the next leader?

**Comment by garious** [comment]
The change is quite subtle. Any way you can capture it in a test? Seems like you'd only need to do this trick for `banking_stage`. The purpose is to wait for writes to the Bank to complete?

File: src/tpu.rs [hunk 1]
```diff
@@ -30,11 +30,11 @@ pub enum TpuMode {
 }
 
 pub struct LeaderServices {
-    fetch_stage: FetchStage,
-    sigverify_stage: SigVerifyStage,
-    banking_stage: BankingStage,
-    cluster_info_vote_listener: ClusterInfoVoteListener,
-    broadcast_service: BroadcastService,
+    fetch_stage: Option<FetchStage>,
```
Comment (mvines) [inline src/tpu.rs]: How about wrapping the entire LeaderServices struct in an Option.   Doing so ought to reduce all the taking and unwrapping, and I think it’s all or nothing anyway — that is we never want to start or stop individual stages

**Comment by carllin** [comment]
@garious, yeah the change was just to guarantee that we have cleanly exited all the leader services before starting the next batch of leader services. Less shared state == fewer ways to shoot ourselves in the foot

File: src/tpu.rs (hunk 1 above)
Comment (carllin) [reply src/tpu.rs]: @mvines, sure thing

**Comment by carllin** [comment]
@garious, there was no particular issue I was seeing, it was more of a precaution. Not entirely sure what a meaningful test would look like, maybe just a test that calls exit and close on the tpu?

**Comment by garious** [comment]
So generally we only merge code without tests if there are no functional changes or if we're deleting code. This *adds* code as a precautionary measure. If we do this too often, working in our codebase will feel like walking in mud.  We won't be able to move quickly because we'll be concerned we've reintroduced a bug in code like this.  If, for example, we moved to what Anatoly's proposed here:

https://github.com/solana-labs/solana/pull/2737/files#diff-52fc9c0c1a302aa2cea9c4aebe074f27

would this code exist?

**Pull request merged**